MiniAn/minian_main.py

```diff
@@ -467,6 +471,116 @@ def load_doric_to_xarray(
+
+    # Estimate a translational shift along the session dimension using the max projection for each dataset.
+    shifts = estimate_motion(concatenated_maxAC, dim = "session").compute().rename("shifts")
+    # Apply Shifts
+    transformed = apply_transform(concatenated_maxAC, shifts).compute().rename("temps_shifted")
+    shiftds = xr.merge([concatenated_maxAC, shifts, transformed])
+
+    window    = shiftds["temps_shifted"].isnull().sum("session")
+    window, _ = xr.broadcast(window, shiftds["temps_shifted"])
+
+    # get ROI footprints ('A') of the base (reference) file
+    ref_footprints = get_footprints(ref_filename, ref_ROIs, ref.coords)
+    file_ref.close()
+
+    # change the unitIds of the current file footprints to number starting from max UnitId in ref file
+    updated_unit_ids = []
+    ref_unit_id_max = (ref_footprints.coords["unit_id"].values).max() + 1
+    for idx, a in enumerate(A.coords["unit_id"]):
+        updated_unit_ids.append(ref_unit_id_max)
+        ref_unit_id_max = ref_unit_id_max + 1
+    A["unit_id"] = updated_unit_ids
+
+    id_dims =  []
+    merged_footprints  = xr.concat([ref_footprints, A], pd.Index(["session1", "session2"], name="session"))
+
+    # Apply shifts to spatial footprint of each session
+    A_shifted = apply_transform(merged_footprints.chunk(dict(height = -1, width = -1)), shiftds["shifts"])
+
+    def set_window(wnd):
+        return wnd == wnd.min()
+    window = xr.apply_ufunc(
+        set_window,
+        window,
+        input_core_dims=[["height", "width"]],
+        output_core_dims=[["height", "width"]],
+        vectorize=True)
+
+    # Calculate centroids of spatial footprints for cells inside a window.
+    cents = calculate_centroids(A_shifted, window)
+                
+    # Calculate pairwise distance between centroids across all pairs of sessions.
+    # To avoid calculating distance between centroids that are very far, a 2d rolling window is applied
+    # to spatial coordinates, & only pairs of centroids within the rolling windows are considered for calculation. default (50, 50).
+    dist = calculate_centroid_distance(cents, "session", id_dims)
+
+    # Threshold centroid distances, keeping only cell pairs with distance less than param_dist.
+    dist_ft = dist[dist["variable", "distance"] < param_dist].copy()
+    dist_ft = group_by_session(dist_ft)
+
+    # Generate mappings
+    mappings = calculate_mapping(dist_ft)
+    mappings_meta = resolve_mapping(mappings)
+    mappings_meta_fill = fill_mapping(mappings_meta, cents)
+    mappings_meta_fill.head()
+
+    # update unitIds of the current file footprints if they have a mapped unitID in ref file
+    for i in range(len(mappings_meta_fill)):
+        if mappings_meta_fill.iloc[i]["group"][0] == ("session1", "session2"):
+            updated_unit_ids = [mappings_meta_fill.iloc[i]["session"]["session1"] if x==mappings_meta_fill.iloc[i]["session"]["session2"] else x for x in updated_unit_ids]
```

Comment:
Please don't write this loop in one line, I hardly can understand what you're doing here. Please rewrite